SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCHandler.cpp

```diff
@@ -0,0 +1,126 @@
+﻿// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Interop/CrossServerRPCHandler.h"
+
+#include "Interop/Connection/SpatialWorkerConnection.h"
+#include "SpatialGDKSettings.h"
+
+DEFINE_LOG_CATEGORY(LogCrossServerRPCHandler);
+
+using namespace SpatialGDK;
+
+CrossServerRPCHandler::CrossServerRPCHandler(ViewCoordinator& InCoordinator, TUniquePtr<RPCExecutorInterface> InRPCExecutor)
+	: Coordinator(InCoordinator)
+	, RPCExecutor(MoveTemp(InRPCExecutor))
+{
+}
+
+void CrossServerRPCHandler::ProcessOps(const float TimeAdvancedS, const TArray<Worker_Op>& WorkerMessages)
+{
+	for (auto& Op : WorkerMessages)
+	{
+		if (Op.op_type == WORKER_OP_TYPE_COMMAND_REQUEST
+			&& Op.op.command_request.request.command_index == SpatialConstants::UNREAL_RPC_ENDPOINT_COMMAND_ID
+			&& Op.op.command_request.request.component_id == SpatialConstants::SERVER_TO_SERVER_COMMAND_ENDPOINT_COMPONENT_ID)
+		{
+			HandleWorkerOp(Op);
+		}
+	}
+
+	ProcessPendingCommandOps();
+}
+
+void CrossServerRPCHandler::ProcessPendingCommandOps()
+{
+	TArray<Worker_EntityId> EmptyRPCQueues;
+	for (auto& CrossServerRPCs : QueuedCrossServerRPCs)
+	{
+		int32 ProcessedRPCs = 0;
+		for (auto& Command : CrossServerRPCs.Value)
+		{
+			if (!TryExecuteCommandRequest(Command))
+			{
+				break;
+			}
+
+			RPCGuidsInFlight.Remove(Command.Payload.UniqueId);
+			++ProcessedRPCs;
+		}
+
+		CrossServerRPCs.Value.RemoveAt(0, ProcessedRPCs);
+		if (CrossServerRPCs.Value.Num() == 0)
+		{
+			EmptyRPCQueues.Add(CrossServerRPCs.Key);
+		}
+	}
+
+	for (auto EntityId : EmptyRPCQueues)
+	{
+		QueuedCrossServerRPCs.Remove(EntityId);
+	}
+}
+
+const TMap<Worker_EntityId_Key, TArray<FCrossServerRPCParams>>& CrossServerRPCHandler::GetQueuedCrossServerRPCs() const
+{
+	return QueuedCrossServerRPCs;
+}
+
+void CrossServerRPCHandler::HandleWorkerOp(const Worker_Op& Op)
+{
+	const Worker_CommandRequestOp& CommandOp = Op.op.command_request;
+	FCrossServerRPCParams Params = RPCExecutor->TryRetrieveCrossServerRPCParams(Op);
+	if (Params.RequestId == -1)
+	{
+		Coordinator.SendEntityCommandFailure(CommandOp.request_id, "Failed to parse cross server RPC", Op.span_id);
+		return;
+	}
+
+	if (RPCGuidsInFlight.Contains(Params.Payload.UniqueId))
+	{
+		// This RPC is already in flight. No need to store it again.
+		UE_LOG(LogCrossServerRPCHandler, Warning, TEXT("RPC is already in flight."));
+		return;
+	}
+
+	if (!QueuedCrossServerRPCs.Contains(CommandOp.entity_id))
+	{
+		// No Command Requests of this type queued so far. Let's try to process it:
+		if (TryExecuteCommandRequest(Params))
+		{
+			return;
+		}
+	}
+
+	// Unable to process command request. Let's queue it up:
+	if (!QueuedCrossServerRPCs.Contains(CommandOp.entity_id))
+	{
+		QueuedCrossServerRPCs.Add(CommandOp.entity_id, TArray<FCrossServerRPCParams>());
+	}
+
+	RPCGuidsInFlight.Add(Params.Payload.UniqueId);
```

Comment:
Does it protect against double sending ? I have the impression that if we receive the command very late, we are still vulnerable to executing the same command that would be sent after a timeout.

Reply:
I think it is impossible to completely prevent double sending with commands. e.g. if an entity moves to another worker, while we re-send an RPC, we will have no way of knowing whether the RPC already got executed.

However, we could keep the RPCs here a bit longer in the set just to make sure that your case is less likely to happen

Reply:
Sure, the case I mentioned was what I supposed to be the motivation for the improvements.